### PR TITLE
Solve remove duplicate

### LINKED ISSUE #21: Challenge_Meduim_Remove duplicates
Remove duplicates ( repeated elements ) in a list.
Bonus: Do it without using the built in function "set()".
Example:
Input: [1, 2, 2, 3, 4, 4, 5]
Output: [1, 2, 3, 4, 5]

### PR DESCRIPTION
**Pull request by safaabuzaid**
---
name: solution review 
about: Review the Remove duplicate function 
---

### Files

- [x] The file name describes the function's behavior
- [x] There is a module docstring in the function file
- [ ] The test file's name matches the function file name -
  `/tests/test_file_name.py`
- [ ] There is a module docstring in the tests file

### Unit Tests

- [x] The test class has a helpful name in PascalCase
- [x] The test class has a docstring
- Every unit test has
  - [x] A helpful name
  - [x] A clear docstring
  - [ ] Only one assertion
  - [x] There is no logic in the unit test
- [x] All tests pass
- [x] There are tests for defensive assertions
- [ ] There are tests for boundary cases

### Function Docstring

- [x] The function's behavior is described
- The function's arguments are described:
  - [x] Type
  - [x] Purpose
  - [x] Other assumptions (eg. if it's a number, what's the expected range?)
- The return value is described
  - [x] Type
  - [x] Other assumptions are documented
- The defensive assertions are documented using `Raises:`
  - [x] Each assumption about an argument is checked with an assertion
  - [x] Each assertion checks for _only one_ assumption about the argument
- [x] Include 3 or more (passing!) doctests

### The Function

- [x] The function's name describes it's behavior
- [x] The function's name matches the file name
- [x] The function has correct type annotations
- [x] The function is not called in the function file

### Do's

- [x] Variable names help to understand the strategy
- [x] Any comments are clear and describe the strategy
- [x] Lines of code are spaced to help show different stages of the strategy

### Don'ts

- [ ] The function's strategy _is not_ described in the documentation
- [ ] Comments explain the _strategy_, **not** the _implementation_
- [ ] The function _does not_ have more comments than code
  - If it does, consider finding a new strategy or a simpler implementation

## Implementation

- [x] The code passes the formatting checks
- [x] The code passes all Ruff linting checks
- [x] The code has no (reasonable) Pylint errors
  - In code review, you can decide when fixing a Pylint error is helpful and
    when it's too restricting.
- [x] Variables are named with snake_case
- [x] Variable names are clear and helpful
- [x] The code follows the strategy as simply as possible
- [x] The implementation is as simple as possible given the strategy
- [x] There are no commented lines of code
- [x] There are no `print` statements anywhere
- [x] The code includes defensive assertions
- [x] Defensive assertions include as little logic as possible

### CODE REVIEW

File: solutions/Remove_Duplicates.py [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+
+Created on 2025-1-4
+@author: Safaa Osman
+"""
+
+def Remove_Duplicates(numbers: list) -> list:
+  """
+  This Function Removes any duplicates elements from the list
+
+  Arguments: list of integeres
```
Comment (Derekkarungani) [inline solutions/Remove_Duplicates.py]: Correct minor typos in the docstring:
"duplicant" → "duplicate"
"integeres" → "integers"
"list of of integers" → "list of integers"

File: solutions/Remove_Duplicates.py [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+Module contents: 
+  - Remove_Duplicates: Remove any duplicant number in the list
+
+Created on 2025-1-4
+@author: Safaa Osman
+"""
+
+def Remove_Duplicates(numbers: list) -> list:
+  """
+  This Function Removes any duplicates elements from the list
```
Comment (Derekkarungani) [inline solutions/Remove_Duplicates.py]: Since the function doesn't explicitly handle non-integer values, it would be helpful to mention it in the docstring.

File: solutions/Remove_Duplicates.py [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+  [1]
+
+  >>> Remove_Duplicates([])
+  [] 
+
+  >>> Remove_Duplicates([5,5,5,5,5,5,5])
+  [5]
+
+  """
+  assert isinstance(numbers,list),"input must be a list"
```
Comment (Derekkarungani) [inline solutions/Remove_Duplicates.py]: The type hint (numbers: list) can be more specific. You can use list[int] to indicate the expected type of elements in the list.

File: solutions/tests/test_Remove_Duplicate.py [hunk 4]
```diff
@@ -0,0 +1,20 @@
+import unittest
+
+from ..Remove_Duplicates import Remove_Duplicates
```
Comment (Derekkarungani) [inline solutions/tests/test_Remove_Duplicate.py:11]: The from ..Remove_Duplicates import Remove_Duplicates statement is incorrect for running tests in most setups. A better one would be from Remove_Duplicates import Remove_Duplicates without the ..

File: solutions/tests/test_Remove_Duplicate.py [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+
+class TestRemoveDuplicates(unittest.TestCase):
+    """Test the Remove_Duplicates function - some tests are buggy!"""
+    
+    def test_empty_list(self):
+        """It should return [] for an empty string"""
```
Comment (Derekkarungani) [inline solutions/tests/test_Remove_Duplicate.py]: The docstring for test_empty_list states: "It should return [] for an empty string," but the input is an empty list, not a string.

File: solutions/tests/test_Remove_Duplicate.py [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+        self.assertEqual(Remove_Duplicates([]), [])
+        
+    def test_no_duplicants(self):
```
Comment (Derekkarungani) [inline solutions/tests/test_Remove_Duplicate.py]: Spelling error on the word duplicants, which should be duplicates.

File: solutions/tests/test_Remove_Duplicate.py [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+        """It should return the same list for list without duplicants"""
+        self.assertEqual(Remove_Duplicates([1,2,3]), [1,2,3])
+
+    
+    def test_not_list(self):
```
Comment (Derekkarungani) [inline solutions/tests/test_Remove_Duplicate.py:25]: You can add a couple more test cases for scenarios such as a list with duplicates, a list with all identical elements.

File: solutions/tests/test_Remove_Duplicate.py [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+        """It should raise AssertionError for non-list input"""
+        with self.assertRaises(AssertionError):
+            Remove_Duplicates("123")
```
Comment (Derekkarungani) [inline solutions/tests/test_Remove_Duplicate.py:28]: Good unittest overall. It's well-structured and makes good use of unittest features like assertEqual and assertRaises.

**Review by Derekkarungani — CHANGES_REQUESTED** [review]
A well thought out use of unittest for testing the Remove_Duplicates function, and the structured approach to test coverage. The implementation looks great for the main purpose of removing duplicates from a list. Just a few minor issues to be addressed.

File: solutions/tests/test_Remove_Duplicate.py (hunk 4 above)
Comment (safaabuzaid) [reply solutions/tests/test_Remove_Duplicate.py:11]: Actually, I tried to remove the (two dots) but I got error, so I think I will keep it.

**Comment by safaabuzaid** [comment]
@Derekkarungani thank you for your comments, I have addressed the feedback, please review again and share with me your thoughts.

**Comment by Derekkarungani** [comment]
@safaabuzaid , looks perfect!
